### PR TITLE
[Improvement](like) Change `like` function to batch call

### PR DESCRIPTION
**Pull request by Gabriel39**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/like.cpp [hunk 1]
```diff
@@ -63,35 +63,171 @@ Status LikeSearchState::clone(LikeSearchState& cloned) {
...
+    return Status::OK();
+}
+
+Status FunctionLikeBase::constant_substring_fn(LikeSearchState* state, const ColumnString& val,
+                                               const StringValue& pattern,
+                                               ColumnUInt8::Container& result) {
+    auto sz = val.size();
+    for (size_t i = 0; i < sz; i++) {
+        if (state->search_string_sv.size == 0) {
+            result[i] = true;
```
Comment (HappenLee) [inline be/src/vec/functions/like.cpp:106]: why here only set one result element and return ？ seems wrong？

File: be/src/vec/functions/like.h [hunk 2]
```diff
@@ -166,11 +169,64 @@ class FunctionLikeBase : public IFunction {
...
                                                  unsigned char* result);
 
-    static Status constant_regex_fn(LikeSearchState* state, const StringValue& val,
-                                    const StringValue& pattern, unsigned char* result);
+    static Status constant_regex_fn(LikeSearchState* state, const ColumnString& val,
+                                    const StringValue& pattern, ColumnUInt8::Container& result);
+
+    static Status regexp_fn(LikeSearchState* state, const ColumnString& val,
+                            const StringValue& pattern, ColumnUInt8::Container& result);
+
```
Comment (HappenLee) [inline be/src/vec/functions/like.h:177]: Add a comment the code with name `fn_predicate` only execute in storage engine？ or a better name ?

File: be/src/vec/functions/like.cpp [hunk 3]
```diff
@@ -63,35 +63,171 @@ Status LikeSearchState::clone(LikeSearchState& cloned) {
...
+    return Status::OK();
+}
+
+Status FunctionLikeBase::constant_substring_fn_predicate(
+        LikeSearchState* state, const PredicateColumnType<TYPE_STRING>& val,
+        const StringValue& pattern, ColumnUInt8::Container& result, uint16_t* sel, size_t sz) {
+    auto data_ptr = reinterpret_cast<const StringRef*>(val.get_data().data());
+    for (size_t i = 0; i < sz; i++) {
+        if (state->search_string_sv.size == 0) {
+            result[i] = true;
```
Comment (HappenLee) [inline be/src/vec/functions/like.cpp:158]: same as upper

Comment (HappenLee) [inline be/src/vec/functions/like.cpp:158]: same as upper

File: be/src/vec/functions/like.cpp [hunk 4]
```diff
@@ -63,35 +63,171 @@ Status LikeSearchState::clone(LikeSearchState& cloned) {
...
+    if (ret != HS_SUCCESS && ret != HS_SCAN_TERMINATED) {
+        return Status::RuntimeError(fmt::format("hyperscan error: {}", ret));
+    }
+
+    return Status::OK();
+}
+
+Status FunctionLikeBase::regexp_fn_scalar(LikeSearchState* state, const StringRef& val,
+                                          const StringValue& pattern, unsigned char* result) {
+    std::string re_pattern(pattern.ptr, pattern.len);
```
Comment (HappenLee) [inline be/src/vec/functions/like.cpp:216]: maybe better use `string_view`

File: be/src/vec/functions/like.cpp [hunk 5]
```diff
@@ -120,29 +256,83 @@ Status FunctionLikeBase::constant_substring_fn_vec_dict(LikeSearchState* state,
...
     }
 
     return Status::OK();
 }
 
-Status FunctionLikeBase::regexp_fn(LikeSearchState* state, const StringValue& val,
-                                   const StringValue& pattern, unsigned char* result) {
+Status FunctionLikeBase::regexp_fn(LikeSearchState* state, const ColumnString& val,
+                                   const StringValue& pattern, ColumnUInt8::Container& result) {
     std::string re_pattern(pattern.ptr, pattern.len);
```
Comment (HappenLee) [inline be/src/vec/functions/like.cpp:278]: string_view

File: be/src/vec/functions/like.cpp [hunk 6]
```diff
@@ -196,22 +386,28 @@ Status FunctionLikeBase::execute_impl(FunctionContext* context, Block& block,
...
+                                          &state->search_state));
     } else {
         const auto pattern_col = block.get_by_position(arguments[1]).column;
 
-        if (const auto* patterns = check_and_get_column<ColumnString>(pattern_col.get())) {
-            RETURN_IF_ERROR(vector_vector(values->get_chars(), values->get_offsets(),
-                                          patterns->get_chars(), patterns->get_offsets(), vec_res,
-                                          state->function, &state->search_state));
+        if (const auto* str_patterns = check_and_get_column<ColumnString>(pattern_col.get())) {
+            if (str_patterns->size() != 1) {
```
Comment (HappenLee) [inline be/src/vec/functions/like.cpp:398]: seems only DCHECK here better?

**Comment by hello-stephen** [comment]
https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221015215615_clickbench_pr_29762.html

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**